src/west/commands/project.py

```diff
@@ -29,7 +29,137 @@
+                            help='''projects (by name or path) to operate on;
+                            defaults to all projects in the manifest''')
+
+    def _all_projects(self):
+        # Get a list of project objects from the manifest.
+        #
+        # If the manifest is malformed, a fatal error occurs and the
+        # command aborts.
+
+        try:
+            return list(Manifest.from_file().projects)
+        except MalformedManifest as m:
+            log.die(m.args[0])
+
+    def _cloned_projects(self, args):
+        # Returns _projects(args, listed_must_be_cloned=True) if a
+        # list of projects was given by the user (i.e., listed
+        # projects are required to be cloned).  If no projects were
+        # listed, returns all cloned projects.
+
+        # This approach avoids redundant _cloned() checks
+        return self._projects(args) if args.projects else \
+            [project for project in self._all_projects() if _cloned(project)]
+
+    def _projects(self, args, listed_must_be_cloned=True,
+                  exclude_manifest=False):
+        # Returns a list of project instances for the projects
+        # requested in 'args' (the command-line arguments), in the
+        # same order that they were listed by the user. If
+        # args.projects is empty, no projects were listed, and all
+        # projects will be returned. If a non-existent project was
+        # listed by the user, an error is raised.
+        #
+        # Before the manifest is parsed, it is validated agains a
+        # pykwalify schema.  An error is raised on validation errors.
+        #
+        # listed_must_be_cloned (default: True):
+        #   If True, an error is raised if an uncloned project was listed. This
+        #   only applies to projects listed explicitly on the command line.
+        #
+        # exclude_manifest (default: False):
+        #   If True, the manifest project will not be included in the returned
+        #   list.
+
+        projects = self._all_projects()
+
+        if exclude_manifest:
+            projects.pop(MANIFEST_PROJECT_INDEX)
+
+        if not args.projects:
+            # No projects specified. Return all projects.
+            return projects
+
+        # Sort the projects by the length of their absolute paths,
+        # with the longest path first. That way, projects within
+        # projects (e.g., for submodules) are tried before their
+        # parent projects, when projects are specified via their path.
+        projects.sort(key=lambda project: len(project.abspath), reverse=True)
+
+        # Listed but missing projects. Used for error reporting.
+        missing_projects = []
+
+        def normalize(path):
+            # Returns a case-normalized canonical absolute version of
+            # 'path', for comparisons. The normcase() is a no-op on
+            # platforms on case-sensitive filesystems.
+            return normcase(realpath(path))
+
+        res = []
```

Comment:
When does this code actually get run? When the class is defined?

Reply:
This particular line? It's a method so it gets run when it gets invoked, same as a function. Did you mean some other code?

Reply:
AH. I think I see the confusion. The normalize() def above is a nested function inside the _projects() method definition.

Note that normalize() gets removed entirely later in the series; this first step is just moving code around.

Reply:
Ok thanks for the explanation